ygot/pathstrings.go

```diff
@@ -0,0 +1,268 @@
+
+	gpath := &gnmipb.Path{}
+	for _, p := range parts {
+		name, kv, err := extractKV(p)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing path %s: %v", path, err)
+		}
+		gpath.Elem = append(gpath.Elem, &gnmipb.PathElem{
+			Name: name,
+			Key:  kv,
+		})
+	}
+	return gpath, nil
+}
+
+// pathStringToElements splits the string s, which represents a gNMI string
+// path into its constituent elements. It does not parse keys, which are left
+// unchanged within the path - but removes escape characters from element
+// names. The path returned omits any leading empty elements when splitting
+// on the / character.
+func pathStringToElements(s string) []string {
+	var parts []string
+	var buf bytes.Buffer
+
+	var inKey, inEscape bool
+
+	for _, ch := range s {
+		switch {
+		case ch == '[' && !inEscape:
+			inKey = true
+		case ch == ']' && !inEscape:
+			inKey = false
+		case ch == '\\' && !inEscape && !inKey:
+			inEscape = true
+			continue
+		case ch == '/' && !inEscape && !inKey:
+			parts = append(parts, buf.String())
+			buf.Reset()
+			continue
+		}
+
+		buf.WriteRune(ch)
+		inEscape = false
+	}
+
+	if buf.Len() != 0 {
+		parts = append(parts, buf.String())
+	}
+
+	if len(parts) > 0 && parts[0] == "" {
+		parts = parts[1:]
+	}
+
+	return parts
+}
+
+// extractKV extracts key value predicates from the input string in. It returns
+// the name of the element, a map keyed by key name with values of the predicates
+// specified. It removes escape characters from keys and values where they are
+// specified.
+func extractKV(in string) (string, map[string]string, error) {
+	var inEscape, inKey, inValue bool
+	var name, currentKey string
+	var buf bytes.Buffer
+	keys := map[string]string{}
+
+	for _, ch := range in {
+		switch {
+		case ch == '[' && !inEscape && !inKey:
+			inKey = true
+			if len(keys) == 0 {
+				if buf.Len() == 0 {
+					return "", nil, errors.New("received a value when the element name was null")
+				}
+				name = buf.String()
+				buf.Reset()
+			}
+			continue
+		case ch == ']' && !inEscape:
+			inKey = false
+			inValue = false
+			if err := addKey(keys, name, currentKey, buf.String()); err != nil {
+				return "", nil, err
+			}
+			buf.Reset()
+			currentKey = ""
+			continue
+		case ch == '\\' && !inEscape:
+			inEscape = true
+			continue
+		case ch == '=' && inKey && !inEscape && !inValue:
+			currentKey = buf.String()
+			buf.Reset()
+			inValue = true
+			continue
+		}
+
+		buf.WriteRune(ch)
+		inEscape = false
+	}
+
+	if len(keys) == 0 {
+		name = buf.String()
+	}
+
+	if len(keys) != 0 && buf.Len() != 0 {
+		// In this case, we have trailing garbage following the key.
+		return "", nil, fmt.Errorf("trailing garbage following keys in element %s, got: %v", name, buf.String())
+	}
+
+	return name, keys, nil
+}
+
+// addKey adds key k with value v to the key's map. The key, value pair is specified
+// to be for an element named e.
+func addKey(keys map[string]string, e, k, v string) error {
```

Comment:
Should this (or the caller) trim whitespace from k and v?

Reply:
It can't trim from `v`, since if we have `[foo= bar]` then this is literally `foo` being set to ` bar`, and hence trimming from there is an issue.

Since spaces are not valid YANG identifiers then yes, we can check for spaces there.